### PR TITLE
feat(cli): Assertion signing keys and verification

### PR DESCRIPTION
**Pull request by elizabethhealy**
Add option to the cli to pass in assertion verification keys, handles assertion signing keys passed in on encrypt

### CODE REVIEW

File: lib/tdf3/src/client/builders.ts [hunk 1]
```diff
@@ -682,6 +682,19 @@ class DecryptParamsBuilder {
     return this;
   }
 
+  /**
+   * Sets the assertion verification keys for the decryption parameters.
+   *
+   * @param {AssertionVerificationKeys} assertionVerificationKeys - An array of assertion configurations to be set.
+   * @returns {DecryptParamsBuilder} The current instance of the EncryptParamsBuilder for method chaining.
+   */
+  withAssertionVerificaitonKeys(
```
Comment (dmihalcik-virtru) [inline lib/tdf3/src/client/builders.ts:691]:
```suggestion
  withAssertionVerificationKeys(
```

File: cli/src/cli.ts [hunk 2]
```diff
@@ -120,11 +122,53 @@ function addParams(client: AnyNanoClient, argv: Partial<mainArgs>) {
...
+  return u;
+}
+
 async function tdf3DecryptParamsFor(argv: Partial<mainArgs>): Promise<DecryptParams> {
   const c = new DecryptParamsBuilder();
   if (argv.noVerifyAssertions) {
     c.withNoVerifyAssertions(true);
   }
+  if (argv.assertionVerificationKeys) {
+    c.withAssertionVerificaitonKeys(
```
Comment (dmihalcik-virtru) [inline cli/src/cli.ts:168]:
```suggestion
    c.withAssertionVerificationKeys(
```

**Review by dmihalcik-virtru — APPROVED** [review]
Minor spelling error, otherwise good

**Comment by elizabethhealy** [comment]
ignore extra commits, im trying to sign the autoformat unverified ones

File: cli/src/cli.ts [hunk 3]
```diff
@@ -22,6 +22,8 @@ import { webcrypto } from 'crypto';
 import * as assertions from '@opentdf/sdk/assertions';
 import { attributeFQNsAsValues } from '@opentdf/sdk/nano';
 import { base64 } from '@opentdf/sdk/encodings';
+import { importPKCS8, importSPKI, KeyLike } from 'jose'; // for RS256
+import { TextEncoder } from 'util'; // for HS256
```
Comment (dmihalcik-virtru) [inline cli/src/cli.ts:26]:
```suggestion
```

This should be available in the global namespace

File: cli/src/cli.ts [hunk 4]
```diff
@@ -134,25 +185,75 @@ async function tdf3DecryptParamsFor(argv: Partial<mainArgs>): Promise<DecryptPar
...
+  }
+
   // if u is null or empty, return an empty array
   if (!u) {
     return [];
   }
   const a = Array.isArray(u) ? u : [u];
-  for (const assertion of a) {
+  for (let i = 0; i < a.length; i++) {
+    const assertion = a[i];
```
Comment (dmihalcik-virtru) [inline cli/src/cli.ts:234]: Why not a `for..of` loop?
```suggestion
  for (const assertion of a) {
```

File: cli/src/cli.ts [hunk 5]
```diff
@@ -134,25 +185,75 @@ async function tdf3DecryptParamsFor(argv: Partial<mainArgs>): Promise<DecryptPar
...
     }
+    if (assertion.signingKey) {
+      const { alg, key } = assertion.signingKey;
+      try {
+        a[i].signingKey.key = await correctAssertionKeys(alg, key);
+      } catch (err) {
+        throw new CLIError(
+          'CRITICAL',
+          `Issue converting assertion key from string: ${err.message}`
+        );
```
Comment (dmihalcik-virtru) [inline cli/src/cli.ts:246]:
```suggestion
          `Issue converting assertion key from string: ${err.message}`,
          err,
        );
```

File: cli/src/cli.ts [hunk 6]
```diff
@@ -134,25 +185,75 @@ async function tdf3DecryptParamsFor(argv: Partial<mainArgs>): Promise<DecryptPar
...
-  for (const assertion of a) {
+  for (let i = 0; i < a.length; i++) {
+    const assertion = a[i];
     if (!assertions.isAssertionConfig(assertion)) {
       throw new CLIError('CRITICAL', `invalid assertion config ${JSON.stringify(assertion)}`);
     }
+    if (assertion.signingKey) {
+      const { alg, key } = assertion.signingKey;
+      try {
+        a[i].signingKey.key = await correctAssertionKeys(alg, key);
```
Comment (dmihalcik-virtru) [inline cli/src/cli.ts:241]:
```suggestion
        assertion.signingKey.key = await correctAssertionKeys(alg, key);
```

File: cli/src/cli.ts [hunk 7]
```diff
@@ -134,25 +185,75 @@ async function tdf3DecryptParamsFor(argv: Partial<mainArgs>): Promise<DecryptPar
...
+  throw new CLIError('CRITICAL', `Unsupported signing key algorithm: ${alg}`); // Handle unsupported algs
+}
+
+async function parseAssertionConfig(s: string): Promise<assertions.AssertionConfig[]> {
+  let u;
+  try {
+    u = JSON.parse(s);
+  } catch (err) {
+    // try as file name:
+    const jsonFile = await openAsBlob(s);
```
Comment (dmihalcik-virtru) [inline cli/src/cli.ts:224]: this probably could use a try/catch to prettifiy the explanation but not required

File: cli/src/cli.ts [hunk 8]
```diff
@@ -120,11 +122,60 @@ function addParams(client: AnyNanoClient, argv: Partial<mainArgs>) {
...
+  }
+  if (typeof u !== 'object' || u === null) {
+    throw new Error('Invalid input: The input must be an object');
+  }
+  // handle both cases of "keys"
+  if (!('Keys' in u && typeof u.Keys === 'object')) {
+    if ('keys' in u && typeof u.keys === 'object') {
+      u.Keys = u.keys;
+    } else {
+      throw new Error('Invalid input: invalid structure of assertionVerificationKeys');
```
Comment (dmihalcik-virtru) [inline cli/src/cli.ts:144]: Maybe should be a CLIError

**Review by dmihalcik-virtru — APPROVED** [review]

**Review by dmihalcik-virtru — APPROVED** [review]

**Pull request closed without merging**